### DISCUSSION TITLE
Typescript server compilation error for `graphql.scalar`

### DISCUSSION THREAD

**Discussion by Shishouille** [0no-co/gql.tada#362] [Q&A] [answered]
Hi!

We are currently doing a migration from graphql codegen to gql.tada, but we stumble accross some, let's say, DX issues.
Let's take the documentation example. If I use a scalar in the most basic way: 
`const SONG = graphql.scalar('Media', 'Song');`
Typescript sends me this error: 
`Type instantiation is excessively deep and possibly infinite.ts(2589)`
But if I use the same scalar without any value, there is no Typescript error. Also if I use the `tsc` command, no error is implemented. 
Life example here:
<img width="426" alt="image" src="https://github.com/user-attachments/assets/e2580122-6f28-4694-ae40-454efaa8f351">
It's not happening for every scalars and sometimes it seems to "disappear" by itself (by reloading the window for instance).
I'm using `gql.tada 1.8.3` with the `"gql.tada/ts-plugin"` in the tsconfig. Do you know what we are doing wrong in that case? 🤔 

If it's not too much, I have another interrogation. I wandered how to deal with all the values of an enum using scalars, if we want an array of all the enum values (let's say for a zod.enum validation for instance), as it won't be updated if the back-end changes (only for new values)? Is there a fix for it, except typing an array with the `ReturnType` of the scalar with all its values?

Thank you for your time!

**Comment by kitten** [comment, marked as answer]
It's hard to tell what causes this, to be honest, without a reproduction. Make sure you're outputting the introspection file in the optimised `.d.ts` format. Especially if it's only happening sometimes then that's a little odd.

I'd recommend to make sure that `graphql.scalar` is not itself also used in recursive types. I can see that you've got it in some kind of array there, but if you're passing that array to, say, Zod, that will just deepen the types for TypeScript and make it more likely that you hit the recursion limit, if there's no boundary between their types and `gql.tada`'s types.

Regarding enum values, there's several threads on this here and on our Discord.
- https://github.com/0no-co/gql.tada/discussions/263
- https://github.com/0no-co/gql.tada/discussions/308
- https://github.com/0no-co/gql.tada/discussions/50

In short, you can pass in custom scalar types for you enums via `initGraphQLTada`, if you _really_ want enums/const-enums to be used for them.

Personally, I'd make two points:
- Enums or listing enum values in your front-end code isn't great. If you duplicate the enum into your client code, you might diverge from the API, especially when both the GraphQL schema and client change and are deployed independently. What TypeScript then sees as exhaustive may not in fact be exhaustive.
- If you enumerate enum values, do keep in mind that that's also subject to the above. The schema is free to add-on more values, and that's generally not considered to be a breaking change for GraphQL schemas, since the client shouldn't have to know about that. That doesn't mean it's bad, but just noting this.

If you don't override the enum type (which is basically the general assumption we make, although we do support overriding this as said before), I'd stick to just using type annotations and just types as often as possible.

I'm not 100% sure why you'd have a Zod schema for example if you've got GraphQL validating and deriving types for a given enum value.

In short though, rather than doing `graphql('YourScalar', 'YOUR_ENUM_VALUE')` I'd stick with just using exact string literal types, since the types are represented just as string literals in `gql.tada` and then type derived types or output types using `typeof ReturnType<graphql.scalar<'YourScalar'>>` and the likes.

It's hard to make any more specific recommendation without knowing the code of course

**Reply by Shishouille** [reply]
Alright, thank you for your answer! I really appreciate how rich it is 😄  If I'm able to reproduce it on a simple app I'll send you the repo.
